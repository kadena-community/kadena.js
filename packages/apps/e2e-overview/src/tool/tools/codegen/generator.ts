import * as path from 'path';
import { CodegenAction, CodegenOptions, CodegenResult, CodegenSession, PlaywrightTestCase } from './types.js';

export class PlaywrightGenerator {
  private static readonly DEFAULT_OPTIONS: Required<CodegenOptions> = {
    outputPath: 'tests',
    testNamePrefix: 'MCP',
    includeComments: true,
  };

  private options: Required<CodegenOptions>;

  constructor(options: CodegenOptions = {}) {
    this.validateOptions(options);
    this.options = { ...PlaywrightGenerator.DEFAULT_OPTIONS, ...options };
  }

  private validateOptions(options: CodegenOptions): void {
    if (options.outputPath && typeof options.outputPath !== 'string') {
      throw new Error('outputPath must be a string');
    }
    if (options.testNamePrefix && typeof options.testNamePrefix !== 'string') {
      throw new Error('testNamePrefix must be a string');
    }
    if (options.includeComments !== undefined && typeof options.includeComments !== 'boolean') {
      throw new Error('includeComments must be a boolean');
    }
  }

  async generateTest(session: CodegenSession): Promise<CodegenResult> {
    if (!session || !Array.isArray(session.actions)) {
      throw new Error('Invalid session data');
    }

    const testCase = this.createTestCase(session);
    const testCode = this.generateTestCode(testCase);
    const filePath = this.getOutputFilePath(session);

    return {
      testCode,
      filePath,
      sessionId: session.id,
    };
  }

  private createTestCase(session: CodegenSession): PlaywrightTestCase {
    const testCase: PlaywrightTestCase = {
      name: `${this.options.testNamePrefix}_${new Date(session.startTime).toISOString().split('T')[0]}`,
      steps: [],
      imports: new Set(['test', 'expect']),
    };

    for (const action of session.actions) {
      const step = this.convertActionToStep(action);
      if (step) {
        testCase.steps.push(step);
      }
    }

    return testCase;
  }

  private convertActionToStep(action: CodegenAction): string | null {
    const { toolName, parameters } = action;

    switch (toolName) {
      case 'playwright_navigate':
        return this.generateNavigateStep(parameters);
      case 'playwright_fill':
        return this.generateFillStep(parameters);
      case 'playwright_click':
        return this.generateClickStep(parameters);
      case 'playwright_screenshot':
        return this.generateScreenshotStep(parameters);
      case 'playwright_expect_response':
        return this.generateExpectResponseStep(parameters);
      case 'playwright_assert_response':
        return this.generateAssertResponseStep(parameters);
      case 'playwright_hover':
        return this.generateHoverStep(parameters);
      case 'playwright_select':
        return this.generateSelectStep(parameters);
      case 'playwright_custom_user_agent':
        return this.generateCustomUserAgentStep(parameters);
      default:
        console.warn(`Unsupported tool: ${toolName}`);
        return null;
    }
  }

  private generateNavigateStep(parameters: Record<string, unknown>): string {
    const { url, waitUntil } = parameters;
    const options = waitUntil ? `, { waitUntil: '${waitUntil}' }` : '';
    return `
    // Navigate to URL
    await page.goto('${url}'${options});`;
  }

  private generateFillStep(parameters: Record<string, unknown>): string {
    const { selector, value } = parameters;
    return `
    // Fill input field
    await page.fill('${selector}', '${value}');`;
  }

  private generateClickStep(parameters: Record<string, unknown>): string {
    const { selector } = parameters;
    return `
    // Click element
    await page.click('${selector}');`;
  }

  private generateScreenshotStep(parameters: Record<string, unknown>): string {
    const { name, fullPage = false, path } = parameters;
    const options = [];
    if (fullPage) options.push('fullPage: true');
    if (path) options.push(`path: '${path}'`);
    
    const optionsStr = options.length > 0 ? `, { ${options.join(', ')} }` : '';
    return `
    // Take screenshot
    await page.screenshot({ path: '${name}.png'${optionsStr} });`;
  }

  private generateExpectResponseStep(parameters: Record<string, unknown>): string {
    const { url, id } = parameters;
    return `
    // Wait for response
    const ${id}Response = page.waitForResponse('${url}');`;
  }

  private generateAssertResponseStep(parameters: Record<string, unknown>): string {
    const { id, value } = parameters;
    const assertion = value 
      ? `\n    const responseText = await ${id}Response.text();\n    expect(responseText).toContain('${value}');`
      : `\n    expect(${id}Response.ok()).toBeTruthy();`;
    return `
    // Assert response${assertion}`;
  }

  private generateHoverStep(parameters: Record<string, unknown>): string {
    const { selector } = parameters;
    return `
    // Hover over element
    await page.hover('${selector}');`;
  }

  private generateSelectStep(parameters: Record<string, unknown>): string {
    const { selector, value } = parameters;
    return `
    // Select option
    await page.selectOption('${selector}', '${value}');`;
  }

  private generateCustomUserAgentStep(parameters: Record<string, unknown>): string {
    const { userAgent } = parameters;
    return `
    // Set custom user agent
    await context.setUserAgent('${userAgent}');`;
  }

  private generateTestCode(testCase: PlaywrightTestCase): string {
    const imports = Array.from(testCase.imports)
      .map(imp => `import { ${imp} } from '@playwright/test';`)
      .join('\n');

    return `
${imports}

test('${testCase.name}', async ({ page, context }) => {
  ${testCase.steps.join('\n')}
});`;
  }

  private getOutputFilePath(session: CodegenSession): string {
    if (!session.id) {
      throw new Error('Session ID is required');
    }

    const sanitizedPrefix = this.options.testNamePrefix.toLowerCase().replace(/[^a-z0-9_]/g, '_');
    const fileName = `${sanitizedPrefix}_${session.id}.spec.ts`;
    return path.resolve(this.options.outputPath, fileName);
  }
} 