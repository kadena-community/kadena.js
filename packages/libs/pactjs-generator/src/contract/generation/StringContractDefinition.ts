import { ILogger } from '../parsing/lexer';
import { Defcap, Defun, Output, parser } from '../parsing/parser';

import { IContractDefinition } from './IContractDefinition';

/**
 * @alpha
 */
export class StringContractDefinition implements IContractDefinition {
  private _raw: Output;
  private _contract: string;
  private _logger: ILogger;

  public constructor({
    contract,
    namespace,
    logger = () => {},
  }: {
    contract: string;
    namespace?: string;
    logger?: ILogger;
  }) {
    this._contract = contract;
    this._logger = logger;
    this._raw = parser(this._contract, this._logger);

    this._setNamespace(namespace);
  }

  private _setNamespace(namespace?: string): void {
    if (namespace === undefined) return;

    Object.keys(this._raw).forEach((module) => {
      this._raw[module].namespace = namespace;
    });
  }

  public getCapabilities(
    moduleName: string,
  ): Record<string, Defcap> | undefined {
    return this._raw[moduleName].defcaps;
  }

  public get modules(): string[] | undefined {
    return Object.keys(this._raw);
  }

  public getMethods(moduleName: string): Record<string, Defun> | undefined {
    return this._raw[moduleName].defuns;
  }

  public getNamespace(moduleName: string): string {
    return this._raw[moduleName].namespace;
  }

  public get modulesWithFunctions(): Output {
    return this._raw;
  }
}
