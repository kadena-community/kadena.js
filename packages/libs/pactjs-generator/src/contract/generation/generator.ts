import type { IFunction, IModule, IType } from '../parsing/pactParser';
import { getModuleFullName } from '../parsing/utils/utils';

import { EOL } from 'os';

const keywordsMap: Record<string, string> = {
  decimal: 'IPactDecimal',
  integer: 'IPactInt',
  string: 'string',
  time: 'Date',
  bool: 'boolean',
  guard: 'PactReference',
  object: 'object',
};

const mapType = (inputType?: string | IType): string => {
  if (inputType === undefined) {
    return 'any';
  }
  if (typeof inputType === 'string') {
    return keywordsMap[inputType] ?? 'any';
  }

  if (typeof inputType === 'object' && inputType.kind === 'module')
    return 'PactReference';
  const isList = inputType.isList ? '[]' : '';
  // TODO: import the schema as interface to return kind instead of any
  const type = keywordsMap[inputType.kind] ?? 'any';
  return `${type}${isList}`;
};

const getFuncCapInterfaceName = (func: IFunction): string => {
  if (
    func.allExtractedCaps === undefined ||
    func.allExtractedCaps.length === 0
  ) {
    return '';
  }

  const prefix = func.kind === 'defpact' ? 'defpact_' : '';

  return `ICapability_${prefix}${func.name.replace(/-/g, '_')}`;
};

const indent = (str: string, depth = 1): string =>
  str
    .split(EOL)
    .map((line) => `${' '.repeat(depth * 2)}${line}`)
    .join(EOL);

const getParameters = (
  list?: Array<{
    name: string;
    type:
      | string
      | {
          kind: string;
          value: string;
        };
  }>,
): string[] => {
  if (!list) return [];
  return list.map((arg) => {
    return `${arg.name.replace(/-/g, '')}: ${mapType(arg.type)}`;
  });
};

function genFunCapsInterface(func: IFunction): string {
  if (
    func.allExtractedCaps === undefined ||
    func.allExtractedCaps.length === 0
  ) {
    return '';
  }
  const interfaceName = getFuncCapInterfaceName(func);

  const cap = func.allExtractedCaps.map((cap) => {
    let capabilityName = 'capabilityName';
    while (
      cap.capability.parameters &&
      cap.capability.parameters.find((p) => p.name === capabilityName)
    ) {
      // make sure we don't have a name collision
      capabilityName = `_${capabilityName}`;
    }

    let parameters = [`${capabilityName}: "${cap.fullModuleName}.${cap.name}"`];
    if (cap.capability.parameters) {
      const args = getParameters(cap.capability.parameters);
      parameters = [...parameters, ...args];
    }
    const comment =
      cap.capability.doc !== undefined
        ? `/**${EOL}* ${cap.capability.doc}${EOL}*/`
        : '';
    const addCap = `(${EOL}${parameters
      .map((line) => indent(line))
      .join(`, ${EOL}`)}): ICap`;
    return { comment, addCap };
  });

  const capStr = cap.map((c) => `${c.comment}${EOL}${c.addCap},`).join(EOL);
  return `interface ${interfaceName} {${EOL}${indent(capStr)}${EOL}}`;
}

const getFunctionType = (func: IFunction): string => {
  const capInterfaceName = getFuncCapInterfaceName(func) || '';
  const comment =
    func.doc !== undefined ? `/**${EOL}* ${func.doc}${EOL}*/${EOL}` : '';

  const parameters = getParameters(func.parameters);
  const lnBreak = parameters.length > 1;
  const nl = lnBreak ? EOL : '';
  const caps = capInterfaceName
    ? `${capInterfaceName} & ICapability_Coin_GAS`
    : 'ICapability_Coin_GAS';
  return indent(
    `${comment}"${func.name}": (${nl}${parameters
      .map((d) => (lnBreak ? indent(d) : d))
      .join(`,${nl}`)}) => string & { capability : ${caps}} `,
  );
};

/**
 * @alpha
 */
export function generateDts(
  moduleFullName: string,
  modules: Record<string, IModule>,
): string {
  const module = modules[moduleFullName];
  if (module === undefined) {
    throw new Error(`Module ${moduleFullName} not found`);
  }

  if (module.functions === undefined) {
    throw new Error(`Module ${moduleFullName} has no functions`);
  }

  const functions = module.functions.filter(({ kind }) => kind === 'defun');
  const defpacts = module.functions.filter(({ kind }) => kind === 'defpact');

  const capsInterfaces =
    module.functions
      ?.map(genFunCapsInterface)
      .filter(Boolean)
      .join(EOL.repeat(2)) || '';

  const dts = `
import type { PactReference } from '@kadena/client';
import type { IPactDecimal, IPactInt, ICap } from '@kadena/types';

interface ICapability_Coin_GAS {
  (name: 'coin.GAS'): ICap;
}
${capsInterfaces ? `${EOL}${capsInterfaces}${EOL}` : ''}
declare module '@kadena/client' {
  interface IPactModules {
    ${module.doc ? `/**${EOL}${indent(module.doc, 2)}${indent(EOL, 2)}*/` : ''}
    "${getModuleFullName(module)}": {
${indent(functions.map(getFunctionType).join(`,${EOL}${EOL}`), 2)}
${
  defpacts.length > 0
    ? `
      "defpact":{
${indent(defpacts.map(getFunctionType).join(`,${EOL}${EOL}`), 3)}
      }`
    : ''
}
    }
  }
}`;
  return dts;
}
