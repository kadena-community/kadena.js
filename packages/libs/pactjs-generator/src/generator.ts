import { PathLike, readFileSync } from 'fs';
import byline from 'byline';
import { Defun, getModuleAndMethods, Output } from './parser';

interface IContractDefinition {
  get modules(): string[] | undefined;
  getMethods(moduleName: string): Record<string, Defun> | undefined;
}

export class FileContractDefinition implements IContractDefinition {
  bylineStream: byline.LineStream | undefined;
  raw: Output | undefined;

  constructor(
    private filePath: PathLike,
    private logger: (msg: any, ...args: any[]) => void = () => {},
  ) {
    this.parse();
  }

  private parse(): void {
    this.raw = getModuleAndMethods(
      readFileSync(this.filePath, 'utf8'),
      this.logger,
    );
  }

  get modules() {
    return this.raw ? Object.keys(this.raw) : undefined;
  }

  getMethods(moduleName: string) {
    return this.raw ? this.raw[moduleName].defuns : undefined;
  }
}

export class StringContractDefinition implements IContractDefinition {
  raw: Output | undefined;
  constructor(
    private contract: string,
    private logger: (msg: any, ...args: any[]) => void = () => {},
  ) {
    this.parse();
  }

  private parse(): void {
    this.raw = getModuleAndMethods(this.contract, this.logger);
  }

  get modules(): string[] | undefined {
    return this.raw ? Object.keys(this.raw) : undefined;
  }
  getMethods(moduleName: string): Record<string, Defun> | undefined {
    return this.raw ? this.raw[moduleName].defuns : undefined;
  }
}

export class PactTypescriptGenerator {
  constructor(...args: IContractDefinition[]) {}
}
