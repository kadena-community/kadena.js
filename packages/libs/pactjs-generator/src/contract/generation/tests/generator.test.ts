import { generateDts } from '../generator';
import { StringContractDefinition } from '../StringContractDefinition';

describe('generator', () => {
  it('creates a typescript definition from a contract', () => {
    const contract: string = `(module coin
      (defun transfer:string (from:string to:string amount:decimal))
    )`;
    const parsedContract = new StringContractDefinition(contract);
    const dTs = generateDts(parsedContract.modulesWithFunctions)
      .get('coin')!
      .split(/[\s\n]/)
      .filter((x) => x !== '')
      .join(' ');
    const expected =
      `import type { ICommandBuilder, IPactCommand } from '@kadena/client';
declare module '@kadena/client' {
  export interface ICapabilities { }
  export interface IPactModules {
    "coin": {
      "transfer": (from: string, to: string, amount: number) => ICommandBuilder<ICapabilities> & IPactCommand
    }
  }
}`
        .split(/[\s\n]/)
        .filter((x) => x !== '')
        .join(' ');
    expect(dTs).toBe(expected);
  });

  it('creates a typescript definition with DEFCAPS from a contract', () => {
    const contract: string = `(module coin
      (defun transfer:string (from:string to:string amount:decimal))
      (defcap GAS ())
      (defcap TRANSFER (sender:string receiver:string amount:decimal))
    )`;
    const parsedContract = new StringContractDefinition(contract);
    const dTs = generateDts(parsedContract.modulesWithFunctions)
      .get('coin')!
      .split(/[\s\n]/)
      .filter((x) => x !== '')
      .join(' ');
    const expected =
      `import type { ICommandBuilder, IPactCommand } from '@kadena/client';
declare module '@kadena/client' {
  export interface ICapabilities {
    "coin.GAS": [ ],
    "coin.TRANSFER": [ sender: string, receiver: string, amount: number ]
  }
  export interface IPactModules {
    "coin": {
      "transfer": (from: string, to: string, amount: number) => ICommandBuilder<ICapabilities> & IPactCommand
    }
  }
}`
        .split(/[\s\n]/)
        .filter((x) => x !== '')
        .join(' ');
    expect(dTs).toBe(expected);
  });

  it('creates a typescript definition without incompatible types', () => {
    const contract: string = `(module module-with-dashes
      (defun transfer:string (from:string to:string amount))
    )`;

    const parsedContract = new StringContractDefinition(contract);
    const dTs = generateDts(parsedContract.modulesWithFunctions)
      .get('module-with-dashes')!
      .split(/[\s\n]/)
      .filter((x) => x !== '')
      .join(' ');

    const expected =
      `import type { ICommandBuilder, IPactCommand } from '@kadena/client';
declare module '@kadena/client' {
  export interface ICapabilities {
  }
  export interface IPactModules {
    "module-with-dashes": {
      "transfer": (from: string, to: string, amount: any) => ICommandBuilder<ICapabilities> & IPactCommand
    }
  }
}`
        .split(/[\s\n]/)
        .filter((x) => x !== '')
        .join(' ');
    expect(dTs).toBe(expected);
  });

  it('creates a typescript definition without incompatible types', () => {
    const contract: string = `(module module-with-dashes
  (defun transfer:string (from:string to:string amount))
  (defcap GAS ())
  (defcap TRANSFER (from:string to:string amount))
)`;

    const parsedContract = new StringContractDefinition(contract);
    generateDts(parsedContract.modulesWithFunctions);
    const contractCaps = parsedContract.getCapabilities('module-with-dashes')!;
    const contractMethods = parsedContract.getMethods('module-with-dashes')!;

    expect(Object.keys(contractCaps)).toEqual(['GAS', 'TRANSFER']);
    expect(Object.keys(contractCaps.TRANSFER.args)).toEqual([
      'from',
      'to',
      'amount',
    ]);
    expect(Object.keys(contractMethods)).toEqual(['transfer']);
    expect(Object.keys(contractMethods.transfer.args)).toEqual([
      'from',
      'to',
      'amount',
    ]);
    expect(parsedContract.modules).toEqual(['module-with-dashes']);
  });

  it('creates a typescript definition with a namespace', () => {
    const contract: string = `(namespace 'free-namespace)
    (module the-free-module
      (defun transfer:string (from:string to:string amount:decimal))
    )`;
    const parsedContract = new StringContractDefinition(contract);
    const dTs = generateDts(parsedContract.modulesWithFunctions)
      .get('the-free-module')!
      .split(/[\s\n]/)
      .filter((x) => x !== '')
      .join(' ');
    const expected =
      `import type { ICommandBuilder, IPactCommand } from '@kadena/client';
declare module '@kadena/client' {
  export interface ICapabilities { }
  export interface IPactModules {
    "free-namespace.the-free-module": {
      "transfer": (from: string, to: string, amount: number) => ICommandBuilder<ICapabilities> & IPactCommand
    }
  }
}`
        .split(/[\s\n]/)
        .filter((x) => x !== '')
        .join(' ');
    expect(dTs).toBe(expected);
  });

  it('creates a typescript definition with DEFCAPS from a namespaced contract', () => {
    const contract: string = `(namespace 'free-namespace)
    (module the-free-module
    (defun transfer:string (from:string to:string amount:decimal))
    (defcap GAS ())
    (defcap TRANSFER (sender:string receiver:string amount:decimal))
  )`;
    const parsedContract = new StringContractDefinition(contract);
    const dTs = generateDts(parsedContract.modulesWithFunctions)
      .get('the-free-module')!
      .split(/[\s\n]/)
      .filter((x) => x !== '')
      .join(' ');
    const expected =
      `import type { ICommandBuilder, IPactCommand } from '@kadena/client';
declare module '@kadena/client' {
  export interface ICapabilities {
    "free-namespace.the-free-module.GAS": [ ],
    "free-namespace.the-free-module.TRANSFER": [ sender: string, receiver: string, amount: number ]
  }
  export interface IPactModules {
    "free-namespace.the-free-module": {
      "transfer": (from: string, to: string, amount: number) => ICommandBuilder<ICapabilities> & IPactCommand
    }
  }
}`
        .split(/[\s\n]/)
        .filter((x) => x !== '')
        .join(' ');
    expect(dTs).toBe(expected);
  });

  it('creates a typescript definition with a custom interface name', () => {
    const contract: string = `(namespace 'free-namespace)
    (module the-free-module
    (defun transfer:string (from:string to:string amount:decimal))
    (defcap GAS ())
    (defcap TRANSFER (sender:string receiver:string amount:decimal)))`;

    const parsedContract = new StringContractDefinition(contract);

    const dTs = generateDts(
      parsedContract.modulesWithFunctions,
      'IMyInterfaceName',
    )
      .get('the-free-module')!
      .split(/[\s\n]/)
      .filter((x) => x !== '')
      .join(' ');

    const expected =
      `import type { ICommandBuilder, IPactCommand } from '@kadena/client';
declare module '@kadena/client' {
  export interface IMyInterfaceName {
    "free-namespace.the-free-module.GAS": [ ],
    "free-namespace.the-free-module.TRANSFER": [ sender: string, receiver: string, amount: number ]
  }
  export interface IPactModules {
    "free-namespace.the-free-module": {
      "transfer": (from: string, to: string, amount: number) => ICommandBuilder<IMyInterfaceName> & IPactCommand
    }
  }
}`
        .split(/[\s\n]/)
        .filter((x) => x !== '')
        .join(' ');

    expect(dTs).toBe(expected);
  });
});
