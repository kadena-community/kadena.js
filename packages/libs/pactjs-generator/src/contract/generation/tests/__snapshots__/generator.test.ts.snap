// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generateDts add defpact functions to defpact property 1`] = `
"
import type { PactReference } from '@kadena/client';
import type { IPactDecimal, IPactInt, ICap } from '@kadena/types';

interface ICapability_Coin_GAS {
  (name: 'coin.GAS'): ICap;
}

declare module '@kadena/client' {
  export interface IPactModules {
    
    "user.test-module": {
    

      "defpact":{
        "test": () => string & { capability : ICapability_Coin_GAS} 
      }
    }
  }
}"
`;

exports[`generateDts adds IPactReference as the arg type if it is a module reference 1`] = `
"
import type { PactReference } from '@kadena/client';
import type { IPactDecimal, IPactInt, ICap } from '@kadena/types';

interface ICapability_Coin_GAS {
  (name: 'coin.GAS'): ICap;
}

interface ICapability_test_func {
  /**
  * this is defcap doc
  */
  (
    capabilityName: "user.test-module.test-cap", 
    name: string): ICap,
}

declare module '@kadena/client' {
  export interface IPactModules {
    /**
    this is module doc    
    */
    "user.test-module": {
      /**
      * this is defun doc
      */
      "test-func": (
        parameterone: PactReference,
        parametertwo: boolean) => string & { capability : ICapability_test_func & ICapability_Coin_GAS} 

    }
  }
}"
`;

exports[`generateDts adds module, capability and function docs to the dts file 1`] = `
"
import type { PactReference } from '@kadena/client';
import type { IPactDecimal, IPactInt, ICap } from '@kadena/types';

interface ICapability_Coin_GAS {
  (name: 'coin.GAS'): ICap;
}

interface ICapability_test_func {
  /**
  * this is defcap doc
  */
  (
    capabilityName: "user.test-module.test-cap", 
    name: string): ICap,
}

declare module '@kadena/client' {
  export interface IPactModules {
    /**
    this is module doc    
    */
    "user.test-module": {
      /**
      * this is defun doc
      */
      "test-func": (
        parameterone: object,
        parametertwo: boolean) => string & { capability : ICapability_test_func & ICapability_Coin_GAS} 

    }
  }
}"
`;

exports[`generateDts adds some _ to capabilityName to make it unique and avoid name collision if the capability function has also an argument exactly with the same name 1`] = `
"
import type { PactReference } from '@kadena/client';
import type { IPactDecimal, IPactInt, ICap } from '@kadena/types';

interface ICapability_Coin_GAS {
  (name: 'coin.GAS'): ICap;
}

interface ICapability_defpact_test_func {
  /**
  * this is defcap doc
  */
  (
    __capabilityName: "user.test-module.test-cap", 
    capabilityName: string, 
    _capabilityName: string): ICap,
}

declare module '@kadena/client' {
  export interface IPactModules {
    /**
    this is module doc    
    */
    "user.test-module": {
    

      "defpact":{
        /**
        * this is defpact doc
        */
        "test-func": (
          parameterone: any,
          parametertwo: any) => string & { capability : ICapability_defpact_test_func & ICapability_Coin_GAS} 
      }
    }
  }
}"
`;

exports[`generateDts does not generate capability interface if the function uses no capabilities 1`] = `
"
import type { PactReference } from '@kadena/client';
import type { IPactDecimal, IPactInt, ICap } from '@kadena/types';

interface ICapability_Coin_GAS {
  (name: 'coin.GAS'): ICap;
}

declare module '@kadena/client' {
  export interface IPactModules {
    
    "user.test-module": {
      "test-func": (
        parameterone: IPactDecimal,
        parametertwo: boolean) => string & { capability : ICapability_Coin_GAS} 

    }
  }
}"
`;

exports[`generateDts function parameter is empty if the defun has no param 1`] = `
"
import type { PactReference } from '@kadena/client';
import type { IPactDecimal, IPactInt, ICap } from '@kadena/types';

interface ICapability_Coin_GAS {
  (name: 'coin.GAS'): ICap;
}

declare module '@kadena/client' {
  export interface IPactModules {
    
    "user.test-module": {
      "test": () => string & { capability : ICapability_Coin_GAS} 

    }
  }
}"
`;

exports[`generateDts prefixes capabulities for defpact with defpact_ 1`] = `
"
import type { PactReference } from '@kadena/client';
import type { IPactDecimal, IPactInt, ICap } from '@kadena/types';

interface ICapability_Coin_GAS {
  (name: 'coin.GAS'): ICap;
}

interface ICapability_defpact_test_func {
  /**
  * this is defcap doc
  */
  (
    capabilityName: "user.test-module.test-cap", 
    name: string): ICap,
}

declare module '@kadena/client' {
  export interface IPactModules {
    /**
    this is module doc    
    */
    "user.test-module": {
    

      "defpact":{
        /**
        * this is defpact doc
        */
        "test-func": (
          parameterone: any,
          parametertwo: any) => string & { capability : ICapability_defpact_test_func & ICapability_Coin_GAS} 
      }
    }
  }
}"
`;

exports[`generateDts return type definition file for a module 1`] = `
"
import type { PactReference } from '@kadena/client';
import type { IPactDecimal, IPactInt, ICap } from '@kadena/types';

interface ICapability_Coin_GAS {
  (name: 'coin.GAS'): ICap;
}

interface ICapability_test_func {
  
  (
    capabilityName: "user.test-module.test-cap", 
    name: string): ICap,
}

declare module '@kadena/client' {
  export interface IPactModules {
    
    "user.test-module": {
      "test-func": (
        parameterone: string,
        parametertwo: boolean) => string & { capability : ICapability_test_func & ICapability_Coin_GAS} 

    }
  }
}"
`;

exports[`generateDts should generate the function also if return type is a list 1`] = `
"
import type { PactReference } from '@kadena/client';
import type { IPactDecimal, IPactInt, ICap } from '@kadena/types';

interface ICapability_Coin_GAS {
  (name: 'coin.GAS'): ICap;
}

declare module '@kadena/client' {
  export interface IPactModules {
    
    "user.test-module": {
      /**
      * Get all events
      */
      "get-events-list": () => string & { capability : ICapability_Coin_GAS} 

    }
  }
}"
`;

exports[`generateDts use any type if function parameters dont have a type 1`] = `
"
import type { PactReference } from '@kadena/client';
import type { IPactDecimal, IPactInt, ICap } from '@kadena/types';

interface ICapability_Coin_GAS {
  (name: 'coin.GAS'): ICap;
}

interface ICapability_test_func {
  
  (
    capabilityName: "user.test-module.test-cap", 
    name: string): ICap,
}

declare module '@kadena/client' {
  export interface IPactModules {
    
    "user.test-module": {
      "test-func": (
        parameterone: any,
        parametertwo: boolean) => string & { capability : ICapability_test_func & ICapability_Coin_GAS} 

    }
  }
}"
`;

exports[`generateDts use object type if function parameters is object{schema} 1`] = `
"
import type { PactReference } from '@kadena/client';
import type { IPactDecimal, IPactInt, ICap } from '@kadena/types';

interface ICapability_Coin_GAS {
  (name: 'coin.GAS'): ICap;
}

interface ICapability_test_func {
  
  (
    capabilityName: "user.test-module.test-cap", 
    name: string): ICap,
}

declare module '@kadena/client' {
  export interface IPactModules {
    
    "user.test-module": {
      "test-func": (
        parameterone: object,
        parametertwo: boolean) => string & { capability : ICapability_test_func & ICapability_Coin_GAS} 

    }
  }
}"
`;

exports[`generateDts uses the property type if there is no mapped value for that 1`] = `
"
import type { PactReference } from '@kadena/client';
import type { IPactDecimal, IPactInt, ICap } from '@kadena/types';

interface ICapability_Coin_GAS {
  (name: 'coin.GAS'): ICap;
}

declare module '@kadena/client' {
  export interface IPactModules {
    
    "user.test-module": {
      "test": (param: any) => string & { capability : ICapability_Coin_GAS} 

    }
  }
}"
`;
